tests/compiletest.rs

```diff
@@ -1,71 +1,432 @@
+                        compare_output(path, actual, expected),
+                }
+            }
+        }
+        eprintln!();
+        eprintln!(
+            "{} tests failed, {} tests passed, {} skipped",
+            failures.len().to_string().red().bold(),
+            (total - failures.len() - skipped).to_string().green(),
+            skipped.to_string().yellow()
+        );
+        std::process::exit(1);
+    }
+    eprintln!();
     eprintln!(
-        "{}",
-        format!("## Running compile-fail tests in {} against miri for target {}", path, target)
-            .green()
-            .bold()
+        "{} tests passed, {} skipped",
+        (total - skipped).to_string().green(),
+        skipped.to_string().yellow()
     );
+}
+
+#[derive(Debug)]
+enum Error {
+    /// Got an invalid exit status for the given mode.
+    ExitStatus(Mode, ExitStatus),
+    PatternNotFound(String),
+    InlinePatternNotFound(String),
+    NoPatternsFound,
+    OutputDiffers {
+        path: PathBuf,
+        actual: String,
+        expected: String,
+    },
+}
+
+type Errors = Vec<Error>;
+
+fn revisions(path: &Path) -> Vec<String> {
+    let content = std::fs::read_to_string(path).unwrap();
+    for line in content.lines() {
+        if let Some(revisions) = line.strip_prefix("// revisions:") {
+            return revisions.trim().split_whitespace().map(|s| s.to_string()).collect();
+        }
+    }
+    vec![String::new()]
+}
+
+fn run_test(
+    path: &Path,
+    target: &str,
+    flags: &[String],
+    mode: Mode,
+    revision: &str,
+) -> (Command, Errors) {
+    // Run miri
+    let mut miri = Command::new(miri_path());
+    miri.args(flags.iter());
+    miri.arg(path);
+    if !revision.is_empty() {
+        miri.arg(format!("--cfg={revision}"));
+    }
+    miri.env("RUSTC_BACKTRACE", "0");
+    extract_env(&mut miri, path);
+    let output = miri.output().expect("could not execute miri");
+    let mut errors = mode.ok(output.status);
+    // Check output files (if any)
+    let revised = |extension: &str| {
+        if revision.is_empty() {
+            extension.to_string()
+        } else {
+            format!("{}.{}", revision, extension)
+        }
+    };
+    let stderr = check_output(&output.stderr, path, &mut errors, revised("stderr"), target);
+    check_output(&output.stdout, path, &mut errors, revised("stdout"), target);
+    let require = match mode {
+        Mode::Pass => false,
+        Mode::Panic => false, // Should we do anything here?
+        Mode::UB => true,
+    };
+    check_annotations(path, &stderr, &mut errors, require, revision);
+    (miri, errors)
+}
+
+fn check_annotations(
+    path: &Path,
+    stderr: &str,
+    errors: &mut Errors,
+    require: bool,
+    revision: &str,
+) {
+    let content = std::fs::read_to_string(path).unwrap();
+    let mut found_annotation = false;
+    let regex =
+        Regex::new("//(\\[(?P<revision>[^\\]]+)\\])?~[\\^|]*\\s*(ERROR|HELP|WARN)?:?(?P<text>.*)")
```

Comment:
This might be more readable with raw strings, avoiding all the `\\`

Reply:
Also regexps don't usually do escaping inside `[...]`, so it should be `[^|]` I think.

Reply:
`[^x]` means all characters except for `x` as `^` is the inverter for these

Reply:
But using raw strings was definitely better

Reply:
Ah right, but I don't think `\` has any effect in `[`. So make it `[|^]` then.